### ISSUE TITLE
Precision of vector rotations operations.

### ISSUE THREAD

**Issue by dewiniaid** [KSP-KOS/KOS#1663]
While writing kOS code to do some of my own orbit prediction code, I discovered that vectors subjected to ANGLEAXIS() and other rotation mechanisms were particularly imprecise.

I suspect that this is due to the widespread use of the float-using `Quaternion` rather than the double-using  `QuaternionD` (which, incidentally enough, is used in `src/kOS/Suffixed/GeoCoordinates.cs`)

My current workaround is:

```
// Equivalent to A + angleaxis(b, theta), but more accurate.
FUNCTION AngleAxis2 {
	PARAMETER a. //x,y,z
	PARAMETER b. //u,v,w
	PARAMETER theta.
	LOCAL d IS a*b.
	LOCAL c IS COS(theta).
	LOCAL s IS SIN(theta).
	RETURN V(
		b:x*d*(1-c) + a:x*c + (-b:z*a:y+b:y*a:z)*s,
		b:y*d*(1-c) + a:y*c + ( b:z*a:x-b:x*a:z)*s,
		b:z*d*(1-c) + a:z*c + (-b:y*a:x+b:x*a:y)*s
	).	
}

**Comment by hvacengi** [comment]
We seem to be a little inconsistent here.  `QuaternionD` is used only twice in the entire project, and that's in `GeoCoordinates.GetTerrainAltitude`.

As far as I'm aware, kOS tries to use double precision values where possible internally, for instance vectors store their parameters as doubles and scalars do all math as if they are doubles.  It is a little unusual that directions are treated differently.  It appears there are 61 instances using `Quaternion`, including the steering manager.

My first thought would be to leave the references in the steering manager alone for the sake of performance overhead.  However, all of the vectors used by steering are `Vector3d` so we are already forcing an implicit conversion every time that we rotate those vectors.  That implicit conversion isn't free, so which is more costly: double precision rotation or converting each vector?

I'd like to try to benchmark using both structs and see if one or the other provides better performance.

**Comment by dewiniaid** [comment]
My wild guess is that casting is slightly faster, but the difference is unlikely to be noticeable on modern hardware (and by modern, I'm playing on an ancient i5-2500).  (I might be wrong, but I believe a double to float cast only needs to truncate the storage... unlike, say, a int to/from float which needs to actually transform values)

In any case, using QuaternionD definitely outperforms implementing the same logic in KS code, which is what I'm doing now to maintain precision.  And the precision loss was noticeable -- I discovered it when my orbit prediction library (which deserves feature request issue of its own) had position vectors that consistently ended in multiples of 1/8.

**Comment by dewiniaid** [comment]
Oh.  In the case of the steering manager, I suspect the precision doesn't matter as much -- though it would be interesting to know what benchmarking shows.

**Comment by Dunbaratu** [comment]
As to the guesses about why single-precision was used, it's because I didn't know `QuaternionD` was a thing.  If looking at the Unity docs it's not mentioned because it's a KSP thing, similar to KSP's double-precision vectors.  I was just following the lead I saw in Nivekk's old code and didn't know there even was a `QuaternionD`.

**Comment by hvacengi** [comment]
Yeah, and they're in a different namespace from `Vector3d`, at least currently.  They're both in the KSPUtil assembly, but `Vector3d` is in the global namespace while `QuaternionD` is in the `UnityEngine` namespace (and doesn't appear to have as many already defined extension methods).  I guess it makes sense that if they did it for vectors they would do it for quaterions too, but I couldn't have told you that was the case before this post.

**Comment by Dunbaratu** [comment]
I suppose it's possible it didn't exist prior to the Unity5 changes KSP made.  I know that it used to be the case that everything KSP did was in global namespace, so having it under a different namespace seems to point to it being a rather new thing... maybe?

At any rate, regardless of why it wasn't used in the first place, It's probably a really good idea to do the refactor to change all our `Quaternion`'s to `QuaternionD`'s.

I'm going to change the labels on the issue to reflect that it's not technically a bug, because it does work as it is, just not precisely enough.

**Comment by dewiniaid** [comment]
Looks like QuaternionD was present in kOS back to at least mid-2014, so
it's definitely not new ;)

**Comment by hvacengi** [comment]
I did a quick and simple replacement of all `Quaternion` instances with `QuaternionD` and appeared to find an issue where at least one method (angle axis if I recall correctly) would error due to a missing underlying (un-managed) method.  I'll try to research this enough to allow us to get it into the v1.0.0 release, but that could be a complication.

**Comment by dewiniaid** [comment]
Well, as a worst case scenario (if AngleAxis is the only broken function), there's the above KOS code to do it without Quaternions, derived from formulae at http://inside.mines.edu/fs_home/gmurray/ArbitraryAxisRotation/

**Comment by sshilovsky** [comment]
Is the following kind of the same error? It doesn't look like this should ever underflow into zero.

set b to v(1e-6, 1e-6, 1e-6). 
print b:normalized.          
V(0, 0, 0)

**Comment by Dunbaratu** [comment]
Grumble grumble.  I just tried my hand at fixing this and ran into what I think @hvacengi was talking about when I hit this:

```
System.MissingMethodException: Cannot find the requested method.
  at (wrapper managed-to-native) UnityEngine.QuaternionD:INTERNAL_CALL_LookRotation (Vector3d&,Vector3d&)
  at UnityEngine.QuaternionD.LookRotation (Vector3d forward, Vector3d upwards) [0x00000] in <filename unknown>:0 
  at kOS.Function.FunctionHeading.Execute (kOS.SharedObjects shared) [0x00000] in <filename unknown>:0 
  at kOS.Function.FunctionBase.Execute (kOS.Safe.SafeSharedObjects shared) [0x00000] in <filename unknown>:0 
  at kOS.Safe.Function.FunctionManager.CallFunction (System.String functionName) [0x00000] in <filename unknown>:0 
  at kOS.Safe.Execution.CPU.CallBuiltinFunction (System.String functionName) [0x00000] in <filename unknown>:0 
  at kOS.Safe.Compilation.OpcodeCall.StaticExecute (ICpu cpu, Boolean direct, System.Object destination, Boolean calledFromKOSDelegateCall) [0x00000] in <filename unknown>:0 
  at kOS.Safe.Compilation.OpcodeCall.Execute (ICpu cpu) [0x00000] in <filename unknown>:0 
  at kOS.Safe.Execution.CPU.ExecuteInstruction (IProgramContext context, Boolean doProfiling) [0x00000] in <filename unknown>:0 

(Filename: C:/buildslave/unity/build/artifacts/generated/common/runtime/UnityEngineDebugBindings.gen.cpp Line: 64)
```
When trying to run this:
```
            var q = UnityEngine.QuaternionD.LookRotation(VesselUtils.GetNorthVector(currentVessel), currentVessel.upAxis);
```
The compiler claims that UnityEngine.QuaternionD.LookRotation(Vector3d, Vector3D) exists when it compiles against the DLL, but then at runtime it does not exist.

**Comment by hvacengi** [comment]
Yeah, because the managed method points to an unmanaged implementation there's a 2nd layer where the method can get broken. The only solution I see is to write our own. Unless Squad has an ability to select that feature when they compile their own code.

I do wonder if the development version of the unity dlls might fix it locally, even though it won't help with the release version.

**Comment by Dunbaratu** [comment]
No idea what "managed" means here.  Is it a euphamism for "made of MSIL barcodes"?

**Comment by Dunbaratu** [comment]
Stupid phone "correcting" correct things.  Byte codes not bar codes.

**Comment by Dunbaratu** [comment]
Okay I did some searching around and found that, yes "managed code" is Microsoft-speak for "we didn't want to call it 'virtual machine code' or 'bytecde' because that would make the fact that we got the idea from Java too obvious, so we made up a new word for that concept."

"Managed code" means it's on the virtual machine using the MSIL pseudo-machine code, and "unmanaged" means it's not, and instead is something that's implemented in the computer's own native architecture.  

So presumably, they used a native implementation of some quaternion operations for speed reasons (maybe to see if the GPU can help?), and that native implementation is missing.  But the C# wrapper *around* the native implementation, that is meant to hand off the call over to the native code, is still present so the problem isn't apparent at compile time.

So... is the native implementation missing because we need to use another one of the DLLs KSP ships with?  Or is it missing because it never got used at all internally at SQUAD and they never noticed it was missing, or is it missing because it is present in something that they compiled against and added into their EXE but didn't make available publicly for modders?

In any case, it looks like we may just have to manually make our own version of them.  At least rotation operations are a well known well documented thing so we should be able to build our own, even if it is going to be a bit slower (not like that matters in kerboscript code).

Another way to implement it is to reduce magnitude of the vectors down into values floats can handle no problem (i.e. normalize them to unit vectors), then perform the operations at that level as floats, and if need be, multiply the resulting small vector's x,y,z values back out to the vector's proper size, as doubles.

**Comment by Dunbaratu** [comment]
update: after a little bit of experimenting around with replacing the un-implemented stuff in QuaternionD with our own utility methods, I keep continually getting more and more un-implemented internal methods and properties found at runtime that I have to re-implement, some of which look *really* basic to the function of the class to me (I can't even do this:  `q.eulerAngles`, where q is a `QuaternionD`, and that looks pretty common to me (as in, how could they not have ever tried to get the Euler angles of a QuaternionD, ever?   I have a hard time believing something this fundamentally basic was left un-implemented because it's hard for me to believe they never had cause to use it themselves.).)

This is pushing me toward one of two possible conclusions:

1 - QuaternionD might have been part of an unfinished refactoring and they didn't get done making all the parts of the class work yet, and so they weren't really using it yet.  Therefore we probably shouldn't be using the class at all.
(or)
2 - Maybe QuaternionD *does* fully work, BUT we're doing something wrong in how we're compiling that's causing us to be missing access to the `extern` methods it needs.

**Comment by Dunbaratu** [comment]
The only reason this issue is being held open is so we can be reminded to periodically check back and see if Squad has added the rest of the operators to `QuaternionD` to allow us to use it.  Until they do, we shouldn't switch everything over to them.

**Comment by lamont-granquist** [comment]
@Dunbaratu Squad should have added all the rest of the APIs to QuaternionD.  I think QuaternionD.LookRotation() was the last one to get converted or something like that.
